hare/broker.go

```diff
@@ -156,34 +187,74 @@ func (b *Broker) eventLoop() {
 	}
 }
 
+func (b *Broker) updateLatestLayer(id InstanceId) {
+	if id <= b.latestLayer { // should expect to update only newer layers
+		b.Error("Tried to update a previous layer expected %v > %v", id, b.latestLayer)
+		return
+	}
+
+	b.latestLayer = id
+}
+
+func (b *Broker) updateSynchronicity(id InstanceId) {
+	if _, ok := b.layerState[id]; ok { // already has result
+		return
+	}
+
+	// not exist means unknown, check & set
+
+	if !b.isNodeSynced() {
+		b.Info("Note: node is not synced. Marking layer %v as invalid", id)
+		b.layerState[id] = invalid // mark invalid
+		return
+	}
+
+	b.layerState[id] = valid // mark valid
+}
+
+func (b *Broker) isSynced(id InstanceId) bool {
+	b.updateSynchronicity(id)
+
+	state, ok := b.layerState[id]
+	if !ok { // not exist means unknown
+		log.Panic("layerState doesn't contain a value after call to updateSynchronicity")
+	}
+
+	return state == valid
+}
+
 // Register a listener to Messages
 // Note: the registering instance is assumed to be started and accepting Messages
-func (b *Broker) Register(id InstanceId) chan *Msg {
-	inbox := make(chan *Msg, InboxCapacity)
-
-	wg := sync.WaitGroup{}
-	wg.Add(1)
+func (b *Broker) Register(id InstanceId) (chan *Msg, error) {
+	res := make(chan chan *Msg, 1)
 	regRequest := func() {
-		if id > b.maxReg {
-			b.maxReg = id
-		}
+		b.updateLatestLayer(id)
 
-		b.outbox[id] = inbox
+		if b.isSynced(id) {
+			b.outbox[id] = make(chan *Msg, InboxCapacity)
 
-		pendingForInstance := b.pending[id]
-		if pendingForInstance != nil {
-			for _, mOut := range pendingForInstance {
-				b.outbox[id] <- mOut
+			pendingForInstance := b.pending[id]
+			if pendingForInstance != nil {
+				for _, mOut := range pendingForInstance {
+					b.outbox[id] <- mOut
+				}
+				delete(b.pending, id)
```

Comment:
you should also delete the relevant entry in `layerState`

Reply:
Opened a bug